[luci] Pass to fuse pre-activation batchnrom

jinevening opened the pull request:
This adds a pass to fuse pre-activation batchnorm

ONE-DCO-1.0-Signed-off-by: Hyukjin Jeong <hj1.jeong@samsung.com>

---
Related to: #4773 
Draft PR: #4769

jinevening commented:
See https://github.sec.samsung.net/AIP/Trinity/issues/141 for more details.

mhs4670go commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,701 @@
+#include <luci/IR/CircleNodes.h>
+#include <luci/Log.h>
+
+#include <iostream>
```
I'm not sure but maybe it is traces of debugging?

jinevening replied:
You're right. I forgot to remove it. Thanks.

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,701 @@
+{
+
+// Check if all elements are non-negative
+bool is_non_negative(luci::CircleConst *node)
```
if `node` is input and we don't change the contents, let's use like `const luci::CircleConst *node`

jinevening replied:
Updated. Thanks.

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,701 @@
+}
+
+// Check if mul is batchnorm mul
+bool is_batchnorm_mul(luci::CircleMul *mul, luci::CircleNode *&pred_node, luci::CircleConst *&gamma)
```
same reasone, can we use `const luci::CircleMul *mul`

jinevening replied:
Updated. Thanks.

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,701 @@
+  auto y = static_cast<luci::CircleNode *>(mul->y());
+
+  luci::CircleNode *pred;
+  luci::CircleConst *constant;
```
@llFreetimell , can these give warning from static code analysis or OK as below codes set them?

jinevening replied:
I've added initialization code

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,701 @@
+}
+
+// Check if add is batchnorm add
+bool is_batchnorm_add(luci::CircleAdd *add, luci::CircleMul *&mul, luci::CircleConst *&beta)
```
same reason, `const luci::CircleAdd *add`

jinevening replied:
Updated. Thanks.

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,701 @@
+    if (loco::succs(filter).size() != 1)
+      return nullptr;
+
+    return conv;
```
(optional) `return dynamic_cast<luci::CircleConv2D *>(node);`

jinevening replied:
Updated. Thanks.

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,701 @@
+  // This is for handling the last part of ResNetV2
+  else if (opcode == luci::CircleOpcode::MEAN)
+  {
+    auto mean = static_cast<luci::CircleMean *>(node);
```
`static_cast` ?

seanshpark replied:
maybe use `loco::must_cast` ?

jinevening replied:
Updated. Thanks.

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,701 @@
+      y->opcode() != luci::CircleOpcode::CIRCLECONST)
+  {
+    pred = y;
+    constant = static_cast<luci::CircleConst *>(x);
```
`static_cast` ? maybe use `loco::must_cast`

jinevening replied:
Updated. Thanks.

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,701 @@
+// Check if mul is batchnorm mul
+bool is_batchnorm_mul(luci::CircleMul *mul, luci::CircleNode *&pred_node, luci::CircleConst *&gamma)
+{
+  auto x = static_cast<luci::CircleNode *>(mul->x());
```
`static_cast` ? maybe use `loco::must_cast`

jinevening replied:
Updated. Thanks.

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,701 @@
+           y->opcode() == luci::CircleOpcode::CIRCLECONST)
+  {
+    pred = x;
+    constant = static_cast<luci::CircleConst *>(y);
```
`static_cast` ? maybe use `loco::must_cast`

jinevening replied:
Updated. Thanks.

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,701 @@
+    if (loco::succs(filter).size() != 1)
+      return nullptr;
+
+    return conv;
```
(optiona) `return dynamic_cast<luci::CircleConv2D *>(node);`

jinevening replied:
Updated. Thanks.

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,701 @@
+
+void update_conv_weights_with_gamma(luci::CircleConv2D *conv, luci::CircleConst *gamma)
+{
+  auto filter = dynamic_cast<luci::CircleConst *>(conv->filter());
```
please use `loco::must_cast`

jinevening replied:
Updated. Thanks.

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,701 @@
+  }
+}
+
+luci::CircleSub *insert_sub(luci::CircleNode *pred, luci::CircleConst *beta)
```
`const luci::CircleNode *pred` and `const luci::CircleConst *beta` ?

jinevening replied:
Changing them to const generates compilation error (because of no interface implementation).

seanshpark commented:
using `static_cast` here seems OK but please use `dynamic_cast` or `loco::must_cast` for consistancy with other codes.

seanshpark commented:
if the argument is only used as `read` please add `const`

seanshpark commented:
this PR code is a bit huge. for better  review, I strongly suggest to split codes. It is very hard to consentrate for a long time reading and understanding the codes.

jinevening commented:
> using static_cast here seems OK but please use dynamic_cast or loco::must_cast for consistancy with other codes.

Ok, I'll follow the convention. I've changed static_cast to loco::must_cast when casting from loco node to circle node. 

> if the argument is only used as read please add const

I've changed them for possible cases.

> this PR code is a bit huge. for better review, I strongly suggest to split codes. It is very hard to concentrate for a long time reading and understanding the codes and then thinking about what may go wrong.

I understand it is hard. I've split the previous PR and uploaded only the first stage of the pass. Now FusePreActivationBatchNormPass.cpp is 244 lines. PTAL :)

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,243 @@
+  luci::CircleConst *constant = nullptr;
+
+  if (x->opcode() == luci::CircleOpcode::CIRCLECONST &&
+      y->opcode() != luci::CircleOpcode::CIRCLECONST)
```
(optional) we can also use tlike this
```
auto x = dynamic_cast<luci::CircleConst*>(mul->x());
auto y = dynamic_cast<luci::CircleConst*>(mul->y());

if (x != nullptr && y == nullptr)
```

jinevening replied:
I've changed the code.

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,243 @@
+/**
+ *  Swap MUL/ADD if they are from batch normalization
+ *
+ *  BEFORE
+ *
+ *           [Mul]  gamma
+ *             |
+ *        [Add + Relu]  beta
+ *  AFTER
+ *
+ *           [Add]  beta/gamma
+ *             |
+ *           [Mul]  gamma
+ *             |
+ *           [Relu]
```
```suggestion
 *  BEFORE
 *           [Mul]  gamma
 *             |
 *        [Add + Relu]  beta
 *
 *  AFTER
 *           [Add]  beta/gamma
 *             |
 *           [Mul]  gamma
 *             |
 *           [Relu]
```

jinevening replied:
Changed. Thanks.

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,243 @@
+    return false;
+
+  if (beta->dtype() != loco::DataType::FLOAT32 || gamma->dtype() != loco::DataType::FLOAT32)
+    throw std::runtime_error("FusePreActivationBatchNormPass only supports fp32 model");
```
maybe `fp32` --> `F32` ?

llFreetimell replied:
Is it correct to die when `gamma->dtype()` is not supported or just returning `false` is correct?

jinevening replied:
Changed to Float32 :)

seanshpark reviewed (approved):
LGTM thank you!

seanshpark commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,240 @@
+  if (add->fusedActivationFunction() != luci::FusedActFunc::RELU)
+    return false;
+
+  if (x->opcode() == luci::CircleOpcode::CIRCLECONST && y->opcode() == luci::CircleOpcode::MUL)
```
(optional) as you changed the pattern to check the node type in `is_batchnorm_mul()`, how about apply here too?

jinevening replied:
This code is more complicated to apply the pattern, because it checks opcode is not only CIRCLECONST but also MUL.

seanshpark replied:
OK :)

llFreetimell commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,240 @@
+    return false;
+
+  if (beta->dtype() != loco::DataType::FLOAT32 || gamma->dtype() != loco::DataType::FLOAT32)
+    throw std::runtime_error("FusePreActivationBatchNormPass only supports Float32 model");
```
Is it correct to die when `dtype` are not supported or is it correct to return `false`?

jinevening replied:
> Is it correct to die when dtype are not supported or is it correct to return false?

I think this is a matter of choice, not correctness :)

In my opinion, a non-float model is a completely wrong input to this pass, so I decided to throw an exception.

llFreetimell reviewed (approved):
LGTM

seanshpark reviewed (approved):
LGTM

jinevening commented:
@mhs4670go PTAL

mhs4670go commented on compiler/luci/pass/include/luci/Pass/FusePreActivationBatchNormPass.h:
```diff
@@ -0,0 +1,42 @@
+
+  std::vector<luci::CircleMul *> _mul_list;
+  std::vector<luci::CircleAdd *> _add_list;
+  std::vector<luci::CircleSub *> _sub_list; // inserted during fusion
```
This variable will be used later?

jinevening replied:
Yes, `_sub_list` will be used later.

mhs4670go commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,240 @@
+  {
+    auto b = beta->at<loco::DataType::FLOAT32>(i);
+    auto g = gamma->at<loco::DataType::FLOAT32>(i);
+    if (b == g)
+    {
+      beta->at<loco::DataType::FLOAT32>(i) = 1;
+    }
+    else
+    {
+      // If g is 0, we use a small value (empirically determined)
+      if (g == 0)
+        g = 1e-10;
+      beta->at<loco::DataType::FLOAT32>(i) = b / g;
+    }
```
```suggestion
      // If g is 0, we use a small value (empirically determined)
      if (g == 0)
        g = 1e-10;
      beta->at<loco::DataType::FLOAT32>(i) = b / g;
    
```
Isn't it possible?

jinevening replied:
If b ==0 and g == 0, beta will be 1 in the previous code and 0 in the suggested code.

Actually, this is a corner case that may not happen in real models. I made it to 1 for simple testing (https://github.com/Samsung/ONE/pull/4769/files#diff-ce6d441c9cf90c4ae46084e6c0b00d6d0031c90b455ca36afa662e7e335156b4R179).

mhs4670go commented on compiler/luci/pass/src/FusePreActivationBatchNormPass.cpp:
```diff
@@ -0,0 +1,240 @@
+  // Step 3. Fuse ADD with the preceding CONV and insert SUB
+
+  // Step 4. Fuse SUB to CONV (SUB -> ADD <- CONV pattern)
+
```
Just curiosity, I'm not sure but if the fusion proceeds sequentially, what if the condition that checks if it can keep going or not is fail? I mean, if Step3 returns false, what can we restore the graph?

jinevening replied:
> I'm not sure but if the fusion proceeds sequentially, what if the condition that checks if it can keep going or not is fail? I mean, if Step3 returns false, how can we restore the graph?

The graph is sequentially changed as much as possible. It is not restored even if the next step fails. This is possible because each step is safe (= produces the same result after applied).

mhs4670go reviewed (approved):
LGTM
